src/lib/homestore/shard_manager.cpp

```diff
@@ -1,15 +1,293 @@
+}
+
+uint64_t HSHomeObject::get_sequence_num_from_shard_id(uint64_t shard_id) {
+    return shard_id & (max_shard_num_in_pg() - 1);
+}
+
+std::string HSHomeObject::serialize_shard(const Shard& shard) const {
+    nlohmann::json j;    
+    j["shard_info"]["shard_id"] = shard.info.id;
+    j["shard_info"]["pg_id"] = shard.info.placement_group;
+    j["shard_info"]["state"] = shard.info.state;
+    j["shard_info"]["created_time"] = shard.info.created_time;
+    j["shard_info"]["modified_time"] = shard.info.last_modified_time;
+    j["shard_info"]["total_capacity"] = shard.info.total_capacity_bytes;
+    j["shard_info"]["available_capacity"] = shard.info.available_capacity_bytes;
+    j["shard_info"]["deleted_capacity"] = shard.info.deleted_capacity_bytes;
+    j["ext_info"]["chunk_id"] = shard.chunk_id;
+    return j.dump();
+}
+
+Shard HSHomeObject::deserialize_shard(const std::string& shard_json_str) const {
+    auto shard_json = nlohmann::json::parse(shard_json_str);  
+    ShardInfo shard_info;
+    shard_info.id = shard_json["shard_info"]["shard_id"].get<shard_id>();
+    shard_info.placement_group = shard_json["shard_info"]["pg_id"].get<pg_id>();
+    shard_info.state = static_cast<ShardInfo::State>(shard_json["shard_info"]["state"].get<int>());
+    shard_info.created_time = shard_json["shard_info"]["created_time"].get<uint64_t>();
+    shard_info.last_modified_time = shard_json["shard_info"]["modified_time"].get<uint64_t>();
+    shard_info.available_capacity_bytes = shard_json["shard_info"]["available_capacity"].get<uint64_t>();
+    shard_info.total_capacity_bytes = shard_json["shard_info"]["total_capacity"].get<uint64_t>();
+    shard_info.deleted_capacity_bytes = shard_json["shard_info"]["deleted_capacity"].get<uint64_t>();
+    auto shard = Shard(shard_info);
+    shard.chunk_id = shard_json["ext_info"]["chunk_id"].get<uint16_t>();
+    return shard;
+}
+
 ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id pg_owner, uint64_t size_bytes) {
-    return folly::makeUnexpected(ShardError::UNKNOWN_PG);
+    auto pg = _get_pg(pg_owner);
+    if (!bool(pg)) {
+        LOGWARN("failed to create shard with non-exist pg [{}]", pg_owner);
+        return folly::makeUnexpected(ShardError::UNKNOWN_PG);
+    }
+    //TODO: will update to ReplDev when ReplDev on HomeStore is ready;
+    auto replica_set_var = _repl_svc->get_replica_set(fmt::format("{}", pg_owner));
+    if (std::holds_alternative< home_replication:: ReplServiceError>(replica_set_var)) {
+        LOGWARN("failed to get replica set instance for pg [{}]", pg_owner);
+        return folly::makeUnexpected(ShardError::UNKNOWN_PG);	
+    }
+
+    auto replica_set = std::get< home_replication::rs_ptr_t >(replica_set_var);
+    if (!replica_set->is_leader()) {
+        LOGWARN("pg [{}] replica set is not a leader, please retry other pg members", pg_owner);
+        return folly::makeUnexpected(ShardError::NOT_LEADER);
+    }
```

Comment:
Oh; shard creation should be able to happen everywhere. We can relax this in M3 I think when we have real replicas.

Reply:
Just to follow up; this _has_ to be the case since by the time we actually call `append_entries` in HomeStore the leadership has changed since it was checked. We can never pre-check leadership...

Reply:
maybe let HS return an error in this case?  Same situation is also true for write_blob request right?

Reply:
Yes; if we want to prevent this we can just turn `forwarding` off in the raft_server; it will reject `append_entries` if it's not currently the leader. But I'm not sure why we care about leadership here...we're basically just doing what the DM would retry anyways by forwarding it.

Reply:
yes, if raft server on follower side support msg forward, then this check will not be necessary. before I do not know this feature is enabled in nuraft.

Reply:
OK yes, only for the optimize_path. DM and GW need to adjust as well. 

Actually I was kind of not sure why  OM/CM team  need that strict regarding sending to leader :